docs/decisions/0004-downstream-effect-events.rst

```diff
@@ -0,0 +1,103 @@
+Implementation of Event Driven Architecture for Exam Downstream Effects
+=======================================================================
+
+Status
+------
+
+Draft (circa Sept 2023)
+
+Context
+-------
+For Milestone 6 of the exams IDA project, we are porting any downstream effects of exam submission and review
+from edx-proctoring to the exams IDA (aka edx-exams). Most of these downstream effects will reflect calls
+that edx-proctoring (which is a plugin) makes to various edx-platform services, such as grades, credits, and certificates.
+
+Since edx-exams is independent, we plan to use the event bus to send info to edx-platform services
+without needing a response as one would in a REST setup.
+
+This is being done to circumvent messy circular dependencies, wherein edx-exams and edx-platform go back and
+forth asking each other for information.
+
+
+Decision
+--------
+We are porting these because they are both essential and easy to translate into events:
+
+  * Grades Override - A call to the grades service to override a grade when an exam attempt is rejected.
+
+  * Undo Grades Override - A call to the grades to undo such an override when marking a rejected attempt as verified.
+
+  * Instructor Delete Attempt - A call to the instructor service to delete an exam attempt.
+
+  * Instructor Complete Attempt - A call to the instructor service that marks the exam as completed.
+
+  * Instructor Is Course Staff - A call to the instructor service to check if a user is course staff.
+
+  * Invalidate Certificate - A call to the certificates service to invalidate a leaner’s edx certificate for a course.
+
+  * Set Credit Requirement Status - A call to the credits service to create or modify a learner’s credit requirement status for an exam.
+
+Here’s what we’re NOT porting (and why):
+
+  * Credit Prerequisite Check:
+
+    * In edx-proctoring, we call the credits service to see if a learner has completed the prerequisites for an exam.
+
+    * Translating this into edx-exams would require a request-response call to be made from edx-exams to edx-platform. This would create an undesired circular dependency, as we already have edx-platform making such calls to edx-exams.
+
+    * Instead, we have decided to implement an endpoint in the credits service that returns the prereq status, which will be called directly from the exams UI.
+
+    * This feature will be developed separately and is considered to be outside of the scope of M6.
+
+  * Name Affirmation:
+
+    * Currently, edx-proctoring calls the name affirmation service in order to match the name that Proctortrack sees on the learner’s ID to the user’s verified name in the LMS database.
+
+    * We do not plan to implement this since It is very likely that 2U will not require the matching of one’s verified name in edx-platform to the names on their IDs in the future.
```

Comment:
Nit.

```suggestion
    * We do not plan to implement this since it is very likely that 2U will not require the matching of one’s verified name in edx-platform to the names on their IDs in the future.
```

Reply:
What consideration has been made to the Open edX community. I don't think Name Affirmation has wide adoption outside of 2U, but is there any concern about making this change on Open edX operators? It could be worth making a post in the Open edX Discuss forum to inform them of this decision, unless you're already planning to circulate this ADR there.

Reply:
I'm not sure if name affirmation for learner IDs has any impact on Open edX operators. What would be the best way to find out if changing this would be a concern?

Reply:
I'd leave 2U intentions out of this entirely. We're mostly leaving this out to keep initial scope light but I wouldn't call it a decision to drop name affirmation support. We can just say it is left out defining this one as we don't believe it to be an essential feature.